ui.apps/src/main/content/jcr_root/apps/authoring-toolkit/depends-on/js/actions/dependsOnUpdateOptionsAction.js

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Action to refresh the option set of a Coral3 Select widget with an OptionProvider.
+ * Used in cases when path to options node is authored via another widget. Change of this widget's value triggers s
+ * new query for options
+ */
+(function (Granite, $, DependsOn) {
+
+    'use strict';
+
+    /**
+     * Finds out whether the two option sets are equal, e.g. to check if the
+     * newly collected options are the same as options already present in the option container
+     *
+     * @param optionSet0 First option set
+     * @param optionSet1 Second option set
+     * @returns {boolean}
+     */
+    function optionsAreSame(optionSet0, optionSet1) {
+        if (!Array.isArray(optionSet0) || !Array.isArray(optionSet1) || optionSet0.length !== optionSet1.length) {
+            return false;
+        }
+        for (let i = 0; i < optionSet0.length; i++) {
+            if (optionSet0[i].value !== optionSet1[i].value) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Replaces existing option set with a new one in DOM
+     *
+     * @param select        Select widget to set options for
+     * @param newOptions    New option set
+     * @param selectedValue The value to mark as selected in the option set
+     */
+    function replaceOptions(select, newOptions, selectedValue) {
+        select.get(0).items.clear();
+        newOptions.forEach(function(item) {
+            select.get(0).items.add({
+                value: item.value,
+                content: {
+                    textContent: item.text
+                },
+                selected: (!selectedValue && !item.value) || selectedValue === item.value
+            });
+        });
+    }
+
+    /**
+     * Registers {@code update-datasource} action in DependsOn registry.
+     * This action performs updating of the option set of the current Select widget
+
+     * @param path      Path value as set in the foreign widget (a path picker) this widget depends on
+     * @param options   Options to be passed with an async HTTP request to the custom datasource (same as described
+     *                  in the {@code CustomDataSourceServlet} javadoc
+     */
+    DependsOn.ActionRegistry.register('update-options', function(path, options) {
```

Comment:
The complexity of the action function is too big

Reply:
Split it into three different functions)